File: internal/telegram/dedupe.go
package telegram

import "sync"

// dedupe remembers the last n update_ids. The fixed-size ring evicts the
// oldest entry so memory stays bounded.
type dedupe struct {
	mu   sync.Mutex
	seen map[int64]struct{}
	ring []int64
	next int
}

func newDedupe(n int) *dedupe {
	return &dedupe{
		seen: make(map[int64]struct{}, n),
		ring: make([]int64, n),
	}
}

// firstSeen records id and reports whether this is its first appearance.
func (d *dedupe) firstSeen(id int64) bool {
	d.mu.Lock()
	defer d.mu.Unlock()
	if _, dup := d.seen[id]; dup {
		return false
	}
	if old := d.ring[d.next]; old != 0 {
		delete(d.seen, old)
	}
	d.ring[d.next] = id
	d.next = (d.next + 1) % len(d.ring)
	d.seen[id] = struct{}{}
	return true
}

File: chic/telegram/dedupe.py
"""Update deduplication — Telegram re-delivers updates it doubts we received."""

from __future__ import annotations


class Dedupe:
    """Remembers the last n update ids in a fixed-size ring (bounded memory)."""

    def __init__(self, n: int = 1024) -> None:
        self._seen: set[int] = set()
        self._ring: list[int] = [0] * n
        self._next = 0

    def first_seen(self, update_id: int) -> bool:
        if update_id in self._seen:
            return False
        old = self._ring[self._next]
        if old != 0:
            self._seen.discard(old)
        self._ring[self._next] = update_id
        self._next = (self._next + 1) % len(self._ring)
        self._seen.add(update_id)
        return True
